### PR TITLE
[python-package] do not copy column-major numpy arrays when creating Dataset

### PR DESCRIPTION
**Pull request by jmoralez**
Changes the logic to flatten the data array to consider its layout:

* If the array is Fortran contiguous, then it's flattened as such and passed to the Dataset constructor indicating that the array is column-major, which avoids copying it to make it row-major.
* If the array is C contiguous, no copy is made (current behavior).
* If the array is not contiguous because it was sliced, then a copy is made to make it C-contiguous (current behavior).

I ran some quick tests with an array of (100k, 50) and the timings to construct the dataset from a C-contiguous and an F-contiguous arrays are roughly the same, so this doesn't introduce extra latency, just avoids the copy.

### CODE REVIEW

**Comment by jmoralez** [comment]
This could also be done for the predict portion, i.e.

https://github.com/microsoft/LightGBM/blob/5151fe85f08e5dccff7d48242dddace51f9c8ede/python-package/lightgbm/basic.py#L1268-L1307

Please let me know if it's ok to include it here or in a separate PR.

File: tests/python_package_test/test_basic.py
Comment (StrikerRUS) [inline tests/python_package_test/test_basic.py:955]: Maybe reuse `rng` fixture?
https://github.com/microsoft/LightGBM/blob/83c0ff3de1925b0e2d4831a9ccb6ffc196aa795b/tests/python_package_test/conftest.py#L5-L7

**Review by StrikerRUS — APPROVED** [review]
LGTM! Thank you for very useful improvement!

I just left a couple of non-blocking comments below for your consideration.

**Comment by StrikerRUS** [comment]
> Please let me know if it's ok to include it here or in a separate PR.

I believe it would be better to have a follow-up PR for prediction after merging this one.

File: tests/python_package_test/test_engine.py
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py:4636]: Maybe also check that `ds_row` and `ds_col` are identical?
https://github.com/microsoft/LightGBM/blob/83c0ff3de1925b0e2d4831a9ccb6ffc196aa795b/python-package/lightgbm/basic.py#L3535-L3536

**Review by jameslamb — COMMENTED** [review]
This looks great to me! I agree with all of @StrikerRUS 's comments, so won't approve until those are addressed. But I don't have any additional comments.

Thanks for working on this, and really nice tests!

File: tests/python_package_test/test_engine.py
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py:4617]: As this test now doesn't include `train()`, I think it should go to `test_basic.py` file.

Comment (jmoralez) [reply tests/python_package_test/test_engine.py]: Sorry. Moved in 38c6786

**Comment by StrikerRUS** [comment]
@jameslamb 
>I agree with all of @StrikerRUS 's comments, so won't approve until those are addressed.

Would you like to take a look after the recent commits?

**Review by jameslamb — APPROVED** [review]
Changes look great to me, thank you!

I looked at the failing Dask tests... they're unrelated to this PR. Documented in #6739.

@jmoralez do you have time in the next few days to investigate that?

**Comment by jmoralez** [comment]
Sure, I'll take a look.

**Pull request merged**
